### PR TITLE
feat: Honeycomb v3.3.0

### PR DESCRIPTION
**Pull request by RobertGemmaJr**
Note that this PR is bringing all of the previous changes into main, `feat-v3.3` is just a holding branch. All of the code has been looked at in other PRs. 3.3 is just about ready to be released!

# What's Changed

## Electron

- The electron forge configuration is moved to `forge.config.js`
- A `preload` script is added to bridge between the Electron and renderer processes
- Mac is now packaged for the `universal` binary
- Electron is updated to its latest version

## Firebase

The default Firestore rules now allow any `participantID` and `studyID` combination to be used at runtime. This is how all of the labs have been using their Firebase so I figured it should stay. The previous rules (the studyID and participantID must be registered somewhere else in the Firestore) are present but commented out

## Experiment

- All experiment code is moved to `src/experiment`
- The `photodiodeGhostBox` is always added as the `prompt` of the trial
- The `pdSpotEncode` function is always called on `on_load` callback of the trial
- All trials are now written in the "jsPsych way" instead of using custom code
- `blocks` are now referred to as `procedures`

## Miscellaneous

- Adds a Content-Security Policy
- `no-unused-vars` is now a warning with eslint
- Any eslint warnings will now fail the github actions
- Adds the Honeycomb hexagon icon to the project in all of its needed forms
- Deletes the unit tests for the utility functions
- Uses inline exports
- Prefer `function(){}` over callback functions
- Rewrites all TODO's to include `@brown-ccv`

### CODE REVIEW

File: .github/workflows/pull_request.yml [hunk 1]
```diff
@@ -1,15 +1,14 @@
 name: Pull Request
-
-# Runs the build and test scripts on PR
 on: pull_request
 
 concurrency:
   group: ${{ github.workflow }}-${{ github.ref }}
   cancel-in-progress: true
 
+# Runs the builds on PR
```
Comment (RobertGemmaJr) [inline .github/workflows/pull_request.yml]:
```suggestion
```

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -57,14 +58,12 @@ jobs:
         run: npm run package:windows
       - name: 📦 Package app installer - Mac
         if: startsWith(matrix.os, 'mac')
+        # TODO @brown-ccv #318: Universal opening with Rosetta?
```
Comment (RobertGemmaJr) [inline .github/workflows/release.yml]:
```suggestion
```

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -73,6 +72,7 @@ jobs:
           echo "version=$(cat package.json | jq -r '.version')" >> $GITHUB_OUTPUT
         shell: bash
 
+      # Upload installers to github release
```
Comment (RobertGemmaJr) [inline .github/workflows/release.yml]:
```suggestion
      # Upload installers to GitHub release
```

File: forge.config.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        setupIcon: "assets/icons/icon.ico",
+        // TODO @brown-ccv #282: Certificates required to prevent anti-virus false positives
+        // certificateFile: "./cert.pfx",
+        // certificatePassword: process.env.CERTIFICATE_PASSWORD,
+      },
+    },
+  ],
+  plugins: [
+    {
+      // https://www.electronforge.io/config/plugins/auto-unpack-natives
```
Comment (RobertGemmaJr) [inline forge.config.js]:
```suggestion
```

File: forge.config.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    },
+    {
+      // Windows Distribution
+      name: "@electron-forge/maker-squirrel",
+      config: {
+        iconUrl: "https://raw.githubusercontent.com/brown-ccv/honeycomb/main/assets/icons/icon.ico",
+        setupIcon: "assets/icons/icon.ico",
+        // TODO @brown-ccv #282: Certificates required to prevent anti-virus false positives
+        // certificateFile: "./cert.pfx",
+        // certificatePassword: process.env.CERTIFICATE_PASSWORD,
```
Comment (RobertGemmaJr) [inline forge.config.js]:
```suggestion
```

File: public/electron/main.js [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+// Early exit when installing on Windows: https://www.electronforge.io/config/makers/squirrel.windows#handling-startup-events
+if (require("electron-squirrel-startup")) app.quit();
+
+// Initialize the logger for any renderer process
+log.initialize({ preload: true });
+
+// TODO @brown-ccv #192: Handle data writing to desktop in a utility process?
+// TODO @brown-ccv #192: Handle video data writing to desktop in a utility process?
+// TODO @brown-ccv #398: Separate log files for each run through?
+// TODO @brown-ccv #429: Use app.getPath('temp') for temporary JSON file
```
Comment (RobertGemmaJr) [inline public/electron/main.js]:
```suggestion
// TODO @brown-ccv #192: Handle data writing to desktop in a utility process
// TODO @brown-ccv #192: Handle video data writing to desktop in a utility process
// TODO @brown-ccv #398: Separate log files for each run through
// TODO @brown-ccv #429: Use app.getPath('temp') for temporary JSON file
```

File: public/electron/main.js [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+}
+
+/**
+ * Receives the Honeycomb config settings and passes them to the CONFIG global in this file
+ * @param {Event} event The Electron renderer event
+ * @param {Object} trigger The metadata for the event code trigger
+ * @param {string} trigger.comName The COM name of the serial port
+ * @param {Object} trigger.eventCodes The list of possible event codes to be triggered
+ * @param {string} productID The name of the product connected to the serial port
+ * @param {string} vendorID The name of the vendor connected to the serial prot
```
Comment (RobertGemmaJr) [inline public/electron/main.js]:
```suggestion
 * @param {string} trigger.productID The name of the product connected to the serial port
 * @param {string} trigger.vendorID The name of the vendor connected to the serial prot
```

File: public/electron/main.js [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+/**
+ * Load the app into the Electron window
+ * In production it loads the local bundle created by the build process
+ */
+function createWindow() {
+  let mainWindow;
+  let appURL;
+
+  console.log(__dirname);
+  console.log(path.join(__dirname, "index.html"));
```
Comment (RobertGemmaJr) [inline public/electron/main.js]:
```suggestion
```

File: public/electron/main.js [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+/********** HELPERS **********/
+
+/**
+ * Load the app into the Electron window
+ * In production it loads the local bundle created by the build process
+ */
+function createWindow() {
+  let mainWindow;
+  let appURL;
+
```
Comment (RobertGemmaJr) [inline public/electron/main.js:271]:
```suggestion
```

File: public/electron/main.js [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+  //   "file",
+  //   (request, callback) => {
+  //     const url = request.url.substr(8);
+  //     callback({ path: path.normalize(`${__dirname}/${url}`) });
+  //   },
+  //   (error) => {
+  //     if (error) console.error("Failed to register protocol");
+  //   }
+  // );
+}
```
Comment (RobertGemmaJr) [inline public/electron/main.js]:
```suggestion
```

File: public/electron/main.js [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+  ipcMain.handle("getCredentials", handleGetCredentials);
+  ipcMain.on("onDataUpdate", handleOnDataUpdate);
+  ipcMain.handle("onFinish", handleOnFinish);
+  ipcMain.on("photodiodeTrigger", handlePhotodiodeTrigger);
+  ipcMain.on("saveVideo", handleSaveVideo);
+  ipcMain.handle("checkSerialPort", handleCheckSerialPort);
+
+  // Setup min files and create the Electron window
+  setupLocalFilesNormalizerProxy();
+  createWindow();
```
Comment (RobertGemmaJr) [inline public/electron/main.js]:
```suggestion
  // Create the Electron window
  createWindow();
```

File: public/index.html [hunk 12]
```diff
@@ -26,18 +39,11 @@
 
     <title>Honeycomb</title>
   </head>
+
```
Comment (RobertGemmaJr) [inline public/index.html]:
```suggestion
```

**Review by broarr — APPROVED** [review]
This is way too big for me to provide a good review. I skimmed the code and didn't see anything glaring

**Pull request merged**